### ISSUE TITLE
[Q] Reverse-engineering `gcloud`

### ISSUE THREAD

**Issue by SamuelMarks** [googleapis/google-cloud-cpp#6578]
Some things are very easy to do with the `gcloud` CLI, like:
```sh
$ export network='default' instance='example-instance' firewall='ssh-http-icmp-fw'
$ gcloud compute networks create "$network"
$ gcloud compute firewall-rules create "$firewall" --network "$network" --allow 'tcp:22,tcp:80,icmp'
$ gcloud compute instances create "$instance" --network "$network" --tags 'http-server' \
                                              --metadata startup-script='#! /bin/bash
  # Installs apache and a custom homepage
  apt update
  apt -y install apache2
  cat <<EOF > /var/www/html/index.html
  <html><body><h1>Hello World</h1>
  <p>This page was created from a start up script.</p>
  </body></html>'
$ # sleep 15s
$ curl $(gcloud compute instances list --filter='name=('"$instance"')' --format='value(EXTERNAL_IP)')
```
(to be exhaustive in commands, tear down with)
```sh
$ gcloud compute instances delete -q "$instance"
$ gcloud compute firewall-rules delete -q "$firewall"
$ gcloud compute networks delete -q "$network"
```

…but it's not clear what the equivalent commands are from the REST API side. Especially considering the HUGE number of options, e.g., at https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert

So I was thinking to just _steal_ whatever `gcloud` does internally.

Running `rg` I found a bunch of these lines:
https://github.com/googleapis/google-auth-library-python/blob/b1a12d2/google/auth/transport/requests.py#L182

Specifically these 5:
```sh
lib/third_party/google/auth/transport/{_aiohttp_requests.py,requests.py,_http_client.py,urllib3.py}
lib/third_party/google_auth_httplib2/__init__.py
```

Below each of them I added `_LOGGER.debug("With body: %s", body)`. But there seems to be some [fancy batching](https://cloud.google.com/compute/docs/api/how-tos/batch) going on because I almost never get that `With body` line 😞 

Now messing with Wireshark to see what I can find, but I'm confident this is a bad rabbit hole to fall down.

Any suggestions?

**Comment by coryan** [comment]
I think the folks in this repository would have no idea how to help with this, sorry.  None of us has any expertise with `gcloud` beyond basic usage.  Maybe try their issue tracker:

https://issuetracker.google.com/issues?q=status:open%20componentid:187143

ask on stack overflow, or maybe join one of the other GCP community forums:

https://cloud.google.com/community

**Comment by SamuelMarks** [comment]
@coryan Yeah this is for the C/C++ implementation of Google Compute that I'm implementing.

No worries, posted here, please share/upvote or what-have-you: https://stackoverflow.com/q/67552029

**Comment by remyabel** [comment]
This does not seem like the right approach. By attempting to "copy" `gcloud`, it seems like you'll be running into opaque implementation details that are not suitable for a library client. I think you are better off looking at the client libraries (at the bottom of the documentation).

**Comment by SamuelMarks** [comment]
@remyabel Yeah I was looking at the nodejs one originally, then I rewrote everything I was doing on `gcloud` so it would be concise enough to share here.

Anyway, it's working now. Turned out `--log-http` was the trick (I was doing `--verbosity=debug`).

🍾 

PS: This is just to get a simple workflow working. The next step is to generate all the interfaces and workflows. Time/clients-appetite permitting.
